### ISSUE TITLE
NEST does not infer Ids on base class

### ISSUE THREAD

**Issue by devarg** [elastic/elasticsearch-net#2089]
**NEST/Elasticsearch.Net version**: 2.3.1

**Elasticsearch version**: 2.3.1

**Description of the problem including expected versus actual behavior**:
NEST does not properly infer Id from attribute if the selected field is on a base class.
**Steps to reproduce**:
1. Create a base class with a public string "Id" parameter
2. Create a class that extends the first one and add the attr. [ElasticsearchType(IdProperty = nameof(Id))]
3. Execute any operation that consumes the Id inferring (I tried with Index and IndexMany)
4. Verify that the Ids are autogenerated because its not recognized.
5. Secondly you may also try adding the Attribute on the base class which also doesn't works.
6. Problem gets fixed when i move the Id attribute to the extending class which may limit the complexity of the entities.

**Comment by russcam** [comment]
Hey @devarg, I've just tried this with NEST 2.3.1 and cannot replicate the issue with the following

```c#
void Main()
{
	var pool = new SingleNodeConnectionPool(new Uri("http://localhost:9200"));
	var defaultIndex = "default-index";
	var connectionSettings = new ConnectionSettings(pool)
			.PrettyJson()
			.DefaultIndex(defaultIndex)
			.DisableDirectStreaming()
			.OnRequestCompleted(response =>
				{
					// log out the request
					if (response.RequestBodyInBytes != null)
					{
						Console.WriteLine(
							$"{response.HttpMethod} {response.Uri} \n" +
							$"{Encoding.UTF8.GetString(response.RequestBodyInBytes)}");
					}
					else
					{
						Console.WriteLine($"{response.HttpMethod} {response.Uri}");
					}

					// log out the response
					if (response.ResponseBodyInBytes != null)
					{
						Console.WriteLine($"Status: {response.HttpStatusCode}\n" +
								 $"{Encoding.UTF8.GetString(response.ResponseBodyInBytes)}\n" +
								 $"{new string('-', 30)}\n");
					}
					else
					{
						Console.WriteLine($"Status: {response.HttpStatusCode}\n" +
								 $"{new string('-', 30)}\n");
					}
				});

	var client = new ElasticClient(connectionSettings);
	
	if (client.IndexExists(defaultIndex).Exists)
		client.DeleteIndex(defaultIndex);
		
	client.CreateIndex(defaultIndex);

	client.Index(new AnotherProject { Id = "foo" });

	client.IndexMany(new[] {
		new AnotherProject { Id = "bar" },
		new AnotherProject { Id = "baz" },
	});
	
	client.Refresh(defaultIndex);
	
	client.Search<AnotherProject>().Dump();
}

[ElasticsearchType(IdProperty = nameof(Id))]
public abstract class Project
{
	public string Id { get; set;}
}

public class AnotherProject : Project
{
}
``` 
This generates the following requests

```javascript
HEAD http://localhost:9200/default-index?pretty=true
Status: 404

------------------------------

PUT http://localhost:9200/default-index?pretty=true 
{}
Status: 200
{
  "acknowledged" : true
}

------------------------------

PUT http://localhost:9200/default-index/anotherproject/foo?pretty=true 
{
  "id": "foo"
}
Status: 201
{
  "_index" : "default-index",
  "_type" : "anotherproject",
  "_id" : "foo",
  "_version" : 1,
  "_shards" : {
    "total" : 2,
    "successful" : 1,
    "failed" : 0
  },
  "created" : true
}

------------------------------

POST http://localhost:9200/_bulk?pretty=true 
{"index":{"_index":"default-index","_type":"anotherproject","_id":"bar"}}
{"id":"bar"}
{"index":{"_index":"default-index","_type":"anotherproject","_id":"baz"}}
{"id":"baz"}

Status: 200
{
  "took" : 34,
  "errors" : false,
  "items" : [ {
    "index" : {
      "_index" : "default-index",
      "_type" : "anotherproject",
      "_id" : "bar",
      "_version" : 1,
      "_shards" : {
        "total" : 2,
        "successful" : 1,
        "failed" : 0
      },
      "status" : 201
    }
  }, {
    "index" : {
      "_index" : "default-index",
      "_type" : "anotherproject",
      "_id" : "baz",
      "_version" : 1,
      "_shards" : {
        "total" : 2,
        "successful" : 1,
        "failed" : 0
      },
      "status" : 201
    }
  } ]
}

------------------------------

POST http://localhost:9200/default-index/_refresh?pretty=true
Status: 200
{
  "_shards" : {
    "total" : 10,
    "successful" : 5,
    "failed" : 0
  }
}

------------------------------

POST http://localhost:9200/default-index/anotherproject/_search?pretty=true 
{}
Status: 200
{
  "took" : 20,
  "timed_out" : false,
  "_shards" : {
    "total" : 5,
    "successful" : 5,
    "failed" : 0
  },
  "hits" : {
    "total" : 3,
    "max_score" : 1.0,
    "hits" : [ {
      "_index" : "default-index",
      "_type" : "anotherproject",
      "_id" : "baz",
      "_score" : 1.0,
      "_source" : {
        "id" : "baz"
      }
    }, {
      "_index" : "default-index",
      "_type" : "anotherproject",
      "_id" : "foo",
      "_score" : 1.0,
      "_source" : {
        "id" : "foo"
      }
    }, {
      "_index" : "default-index",
      "_type" : "anotherproject",
      "_id" : "bar",
      "_score" : 1.0,
      "_source" : {
        "id" : "bar"
      }
    } ]
  }
}

------------------------------
``` 

This also works as expected when the attribute is on the base type. 

Would you mind sharing the code that you're using?

**Comment by devarg** [comment]
Sorry for the delay and btw thanks for the instant response.
I'll close this one as it was way too obvious to see it. I'll post my code anyway to tell the mistake I've made here.

I had this base class looking like this (in a different assembly from the client factory)

`    [ElasticsearchType(IdProperty = nameof(Id))]
    public class User: BaseEntity
    {
        public string Name { get; set; }
        public string LastName { get; set; }
        public DateTime Dob { get; set; }
    }`

And also a class like this extending it:

[ElasticsearchType(IdProperty = nameof(Id))]
`public class User: BaseEntity
    {
        public string Name { get; set; }
        public string LastName { get; set; }
        public DateTime Dob { get; set; }
    }`

i execute the call this way

(Where items is a list of 'User' with property Id)
`_elastic.GetClient().IndexMany(items)`

The problem here was that Id on base class was missing  {get; set;} , when I tried to move the field Id to User class I indeed wrote it right with the get/set. And even worst when I re-added Id field to base class (to check up if I was making a mistake) ¡I wrote it wrong again! So evidence matched perfectly as a bug that didn't exist. And even worst, i was extending many classes from this base class, so it behaved "wrong". I know that the mistake was mine, and 'Ids inferring' section on NEST docs states that infers properties (Not fields), but I think it would be nice to remind it to learners ;)

Thanks for your quick and good work on validating this.
